### DISCUSSION TITLE
Identify windows by title using regular expressions

### DISCUSSION THREAD

**Discussion by joehewitt** [rxhanson/RectanglePro-Community#501] [Ideas]
Love your app in many ways. One way it fails me is it can't identify specific windows by their content. I have several VS Code windows open and want to save and restore each to specific locations. Rectangle Pro seems unable to distinguish one VS Code window from another, so when it does the layout it often assigns positions to the wrong windows. My suggestion - in the App Layout settings panel, allow users to enter a pattern (regex or simpler) to identify each window by its title. This would allow me to position all my VS Code windows exactly where I want them.

**Comment by rxhanson** [comment]
Thanks for the feedback. This is a bug, and I have it slated for the next release. It should be keeping track of the title and laying out windows accordingly.

**Comment by joehewitt** [comment]
That's great, however I don't think just comparing the full title would work for me because VS Code changes the title as I move from file to file. It is the portion of the title that describes the workspace which is constant for the window.  That's why I suggested an app-specific regular expression comparison. Pretty much all tab-based windows are going to change the title when you switch tabs, so your proposed solution would only be able to restore the window layout if the same tab happened to be open as when the layout was saved.

**Comment by ghost** [comment]
This sounds similar to the problem I have with Splashtop (remote support software).

I can get the main Splashtop windows to sticky where I want it to be but when I then connect to a remote computer the window opens unstuck.

Recognising the windows title as "Splashtop Connect" *should* keep them all in order.

**Comment by Pwntastic** [comment]
Seconding this, I usually have a number of chrome and chrome devtools windows open across two monitors. I want the devtools windows to always be in one position and the non-devtools windows in another.

There doesn't currently seem to be a way to set this up. The titles of every window vary, but the devtools windows always start with `DevTools`, so in an ideal scenario I'd have a regex to just match chrome windows with DevTools in the title and put them in one spot, and another position for all other chrome windows, maybe with just a catchall regex.

**Comment by Gamrix** [comment]
Yes, I would also like this feature. 

My current set up is a dual monitor setup. I have 4 different repos open in 4 separate windows for Cursor. I would want two specific ones to be on one monitor and the other two on the other monitor. As the repo names are in the title bar for the apps, a simple regex or even substring match would do the trick.

**Comment by jamescamping** [comment]
I was just about to post a duplicate discussion around this topic. Yes, being able to do a regex for any window title and have it be placed using the App Layout feature would be amazing. In addition to other uses listed, having this for Outlook calendar in one spot and my inbox in another spot would be very useful. I also use this across 2 monitors.

Happy to help beta test or help with creating this feature.

Thanks for creating a great application!

**Comment by rxhanson** [comment]
Regex can lead to a bit of ambiguity and complexity. It sounds like substring matching should be sufficient for the majority of cases?

**Reply by jamescamping** [reply]
Yeah I think that would work

**Comment by jamescamping** [comment]
Don't want to be a pest. but any ideas on when this might get added to the roadmap? Completely understand if there are other larger priorities or features to be implemented.

**Reply by rxhanson** [reply]
I appreciate the check-in. I have some other tasks for App Layouts that I'm working on for this next release, and I'll see if I can roll this in at the same time.

**Comment by CharlesARoy** [comment]
A few App Layout-related comments, questions, and suggestions.

1. I just updated Rectangle Pro, and it seems like App Layouts changed somewhat. Previously, as far as I can recall, I could only specify layouts on a per-app basis. For example, my old layout setup specified iTerm2, Chrome, and Finder (once each), and when I ran it, _all windows_ from each program got moved to the chosen display. After the update, if I only have one occurrence of a program listed in a layout, it will only move one window. So, if I want to move N windows, I have to have N occurrences of the app listed in the App Layout. Does that sound right?

2. Assuming 1. is correct, there's seemingly no way to know which occurrence of a program listed in a layout corresponds to which window of that program. For example, here is one layout:
<img width="414" height="318" alt="Image on 2025-09-11 11 24 29 AM" src="https://github.com/user-attachments/assets/be6cee0e-c5cb-4742-a782-bd34c8c90138" />

As you can see, each of the occurrences of Chrome (and iTerm2) in the layout is nameless, so I have no idea which occurrence maps to which window. Hopefully, the addition of Regex patterns to identify specific windows will be reflected in the App Layout visual for easy distinction. I'd potentially also suggest making it possible to give a custom name to each item in the layout for easier management.

3. For Chrome, as far as I can tell, the window title is taken from the currently active tab:
<img width="512" height="139" alt="Image on 2025-09-11 11 27 29 AM" src="https://github.com/user-attachments/assets/3299d3de-ccdb-4eb7-8949-d0fc089947da" />


This could make it challenging to properly target a window using a Regex pattern. That said, Chrome has an option to name Windows — a feature that I frequently use for long-lived windows (which are also the ones I typically want to target in a layout). Hopefully, the Regex (or substring) setup you are working on will also work for these titles.

<img width="1087" height="450" alt="Image on 2025-09-11 11 15 36 AM" src="https://github.com/user-attachments/assets/eab31727-f3b5-4d0f-adec-f7fd31ec306f" />

Otherwise, love the app and am looking forward to having more control over layouts!

**Reply by rxhanson** [reply]
Thanks for the detailed feedback.

> I just updated Rectangle Pro, and it seems like App Layouts changed somewhat. Previously, as far as I can recall, I could only specify layouts on a per-app basis. For example, my old layout setup specified iTerm2, Chrome, and Finder (once each), and when I ran it, all windows from each program got moved to the chosen display. After the update, if I only have one occurrence of a program listed in a layout, it will only move one window. So, if I want to move N windows, I have to have N occurrences of the app listed in the App Layout. Does that sound right?

There's a checkbox on the config for the layout: "Apply layout to all app windows", and this will bring back the behavior of applying a layout to all of the windows. Let me know if for some reason that doesn't get what you're looking for.

> Assuming 1. is correct, there's seemingly no way to know which occurrence of a program listed in a layout corresponds to which window of that program. As you can see, each of the occurrences of Chrome (and iTerm2) in the layout is nameless, so I have no idea which occurrence maps to which window. Hopefully, the addition of Regex patterns to identify specific windows will be reflected in the App Layout visual for easy distinction. I'd potentially also suggest making it possible to give a custom name to each item in the layout for easier management.

Indeed, this should be resolved and communicated properly once I get configurable window title matching out the door. Not having visibility on the window matching is an unfortunate limitation with the way it currently works, and I'm planning to address this in the next release. I'll try things out with Chrome and make sure it will work for it.

**Comment by jamescamping** [comment]
@rxhanson 
Did you ship this feature and not tell anyone?!?
<img width="313" height="317" alt="image" src="https://github.com/user-attachments/assets/a2319cfb-6fcb-4e28-9c32-60c80a2ed0f0" />
So far its working for me, thank you for implementing it!

**Reply by rxhanson** [reply]
Thanks for finding it! I have a long list of issues that I need to close out, and this was one of them :)
